kvserver: allow lease extensions from followers in propBuf

erikgrinaker opened the pull request:
The Raft proposal buffer currently rejects lease acquisition proposals
from followers if there is an eligible leader, to avoid interfering with
it trying to claim the lease. However, this logic unintentionally
rejected lease extension proposals from followers as well, which could
cause a leaseholder to lose its lease.

This patch changes the proposal buffer to allow lease extension
proposals from followers for extension-based leases as long as they
still hold a valid lease.

Resolves #59179. Follow-up to #55148.

Release note: None

erikgrinaker committed:
kvserver: allow lease extensions from followers in propBuf

The Raft proposal buffer currently rejects lease acquisition proposals
from followers if there is an eligible leader, to avoid interfering with
it trying to claim the lease. However, this logic unintentionally
rejected lease extension proposals from followers as well, which could
cause a leaseholder to lose its lease.

This patch changes the proposal buffer to allow lease extension
proposals from followers for extension-based leases as long as they
still hold a valid lease.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61982)
<!-- Reviewable:end -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @nvanbenschoten, and @tbg)

___
*[pkg/roachpb/api.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVlf7r6CnBAHRSWFNkZ:-MVlf7r6CnBAHRSWFNk_:b-u88xge) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/roachpb/api.go#L281)):*
> ```Go
> func (rlr *RequestLeaseRequest) IsExtension() bool {
> 	return rlr.PrevLease.Replica.StoreID == rlr.Lease.Replica.StoreID &&
> 		rlr.PrevLease.Equivalent(rlr.Lease)
> ```

I had to add `Equivalent()` as a condition here, such that a lease request in a new epoch is not considered an extension of a lease in a previous epoch. Not sure if this makes sense or what the implications are, would appreciate some extra eyes on it.

This was added to make [`TestRequestsOnLaggingReplica`](https://github.com/cockroachdb/cockroach/blob/8560678a4d2e51159bb0ca6c1be54ca2672c33be/pkg/kv/kvserver/client_raft_test.go#L1064) pass -- otherwise, when a partitioned leader rejoined the cluster in a later epoch it would consider its lease request to be an extension of its previous lease (in the previous epoch), causing a request to the old leader to block rather than redirect to the new leader. I believe this scenario was the entire motivation for rejecting lease acquisition proposals in the first place.

`Equivalent()` does not check whether expiration-based leases have expired (only epoch-based leases), which I think could cause the same problem for system ranges. We could e.g. check it against the current time, or even just compare with the current local lease in `propBuf.leaseStatus` to check whether the lease extension proposal is equivalent and still valid.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/roachpb/api.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVlf7r6CnBAHRSWFNkZ:-MVq86rND3yojGhKpd1q:b5xrq9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/roachpb/api.go#L281)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I had to add `Equivalent()` as a condition here, such that a lease request in a new epoch is not considered an extension of a lease in a previous epoch. Not sure if this makes sense or what the implications are, would appreciate some extra eyes on it.

This was added to make [`TestRequestsOnLaggingReplica`](https://github.com/cockroachdb/cockroach/blob/8560678a4d2e51159bb0ca6c1be54ca2672c33be/pkg/kv/kvserver/client_raft_test.go#L1064) pass -- otherwise, when a partitioned leader rejoined the cluster in a later epoch it would consider its lease request to be an extension of its previous lease (in the previous epoch), causing a request to the old leader to block rather than redirect to the new leader. I believe this scenario was the entire motivation for rejecting lease acquisition proposals in the first place.

`Equivalent()` does not check whether expiration-based leases have expired (only epoch-based leases), which I think could cause the same problem for system ranges. We could e.g. check it against the current time, or even just compare with the current local lease in `propBuf.leaseStatus` to check whether the lease extension proposal is equivalent and still valid.
</blockquote></details>

This makes sense, and you have a good point regarding expiration-based leases. It almost seems as though we should include an epoch even in expiration-based leases, unfortunately this isn't going to work since a zero epoch is what encodes a lease as expiration-based:

https://github.com/cockroachdb/cockroach/blob/e924d91081fdc8c66650cf8654b2008d01a8eb41/pkg/roachpb/data.go#L1902-L1904

Comparing against anything from the local replica state isn't going to help us with the problem of requesting a lease based on a stale lease (& leaving requests waiting for our replica to catch up), so it would have to be a timestamp-based route that we're taking. This is mildly annoying as well, as there's no obvious way to distinguish a lease extension (after the range has been dormant for a while) from the kind of extension we want to block. @nvanbenschoten do you have an idea? Maybe adding an auxiliary epoch field that's only used for expiration-based leases? I don't like the look of that but it seems better than trying to use timestamps.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
Please make the commit message explicit about the fact that the change is only about expiration-based leases (right?)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-:-MVqhAlO9j-Wp8I-L2xF:b-k6s639)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 581 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVqcZ_E2DhhAre53RJO:-MVqcZ_E2DhhAre53RJP:b-1p5dn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf.go#L581)):*
> ```Go
> 			// also send lease extensions for an existing leaseholder.
> 			if isExtension {
> 				log.VEventf(ctx, 4, "proposing lease extension even though we're not the leader")
> ```

nit: `4` is to high, generally only used for extremely verbose messages in loops. Use the more usual `2`.
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVqhM9V9XC90lmRHM4_:-MVqhM9V9XC90lmRHM4a:baw060k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf_test.go#L269)):*
> ```Go
> 		var data []byte
> 		if leaseReq {
> 			pd, data = pc.newLeaseProposal(roachpb.Lease{}, false)
> ```

`false /* newLeaseExtension */` to be in compliance with the style guide
But much better to give them names:
```
type leaseOpt bool

const (
brandNewLease leaseOpt = false
extension = true
)
```
___
*[pkg/roachpb/api.go, line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVqcs1e93ZqJ6aMHhC8:-MVqcs1e93ZqJ6aMHhC9:b-mt0qdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/roachpb/api.go#L278)):*
> ```Go
> 
> // IsExtension returns whether the lease request is an extension of an existing
> // lease or whether it is claiming the lease away from a previous owner.
> ```

nit: s/or whether it is claiming the lease away from a previous owner//

The disjunction is confusing because it's not clear if it refers to the cases where `IsExtension` returns `true` or if it refers to the distinction between true or false. Consider clarifying, or better yet just remove the latter part because it's obvious enough.
___
*[pkg/roachpb/api.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVlf7r6CnBAHRSWFNkZ:-MVqeXlX62ycJMCUZ-Tm:bak56ju) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/roachpb/api.go#L281)):*
> Equivalent() does not check whether expiration-based leases have expired (only epoch-based leases), which I think could cause the same problem for system ranges. We could e.g. check it against the current time, or even just compare with the current local lease in propBuf.leaseStatus to check whether the lease extension proposal is equivalent and still valid.

`Equivalent`doesn't check whether either type of lease has expired - it doesn't even have access to a clock.
https://github.com/cockroachdb/cockroach/blob/e924d91081fdc8c66650cf8654b2008d01a8eb41/pkg/roachpb/data.go#L1927
But otherwise I think you're right - I think we should check whether the current lease is valid. That's what it should mean for a "lease" to be an extension. One can't extend a lease it no longer has. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @erikgrinaker)

___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVqhM9V9XC90lmRHM4_:-MVtZAOM-Sev_S85QXxi:bgwmgw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf_test.go#L269)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`false /* newLeaseExtension */` to be in compliance with the style guide
But much better to give them names:
```
type leaseOpt bool

const (
brandNewLease leaseOpt = false
extension = true
)
```
</blockquote></details>

+1 for `false /* newLeaseExtension */`. Please don't introduce an enum for a test helper function that's called 4 times, all in the same file.
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 591 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVtZpy57xEQrj9OYXSx:-MVtZpyCAtIbMwDyEKSk:b2skh3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf_test.go#L591)):*
> ```Go
> 		},
> 		{
> 			name:  "follower, lease extension to known eligible leader",
> ```

This name seems a little off. The follower is the one who holds the lease and is performing an extension, right? So should it be `follower, lease extension despite known eligible leader` or something along those lines?
___
*[pkg/roachpb/api.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVlf7r6CnBAHRSWFNkZ:-MVtaIs_AkiEd65bbvMm:b-4e3ien) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/roachpb/api.go#L281)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> Equivalent() does not check whether expiration-based leases have expired (only epoch-based leases), which I think could cause the same problem for system ranges. We could e.g. check it against the current time, or even just compare with the current local lease in propBuf.leaseStatus to check whether the lease extension proposal is equivalent and still valid.

`Equivalent`doesn't check whether either type of lease has expired - it doesn't even have access to a clock.
https://github.com/cockroachdb/cockroach/blob/e924d91081fdc8c66650cf8654b2008d01a8eb41/pkg/roachpb/data.go#L1927
But otherwise I think you're right - I think we should check whether the current lease is valid. That's what it should mean for a "lease" to be an extension. One can't extend a lease it no longer has. 


</blockquote></details>

This is an interesting dilemma, as it strikes at the question of what constitutes a lease transfer and what lease requests are safe from getting wedged. I don't think the two sets are equivalent.

We consider a lease "extension" at different parts of the code to be one that [does not change the lease's owner](https://github.com/cockroachdb/cockroach/blob/ca60935df431525717592574475f34b5ed422ea9/pkg/kv/kvserver/batcheval/cmd_lease_request.go#L79). We make no attempt to say that the existing lease is still valid - if it is not and has already been replaced, the lease extension will fail.

An argument could be made that a lease that is replaced by a new lease with the same owner but a new sequence is not a true "extension", because writes proposed under the old lease will be rejected if reordered and applying after the new lease. This is the case during a node restart, where the node's `minLeaseProposedTS` is bumped, triggering [this logic](https://github.com/cockroachdb/cockroach/blob/ca60935df431525717592574475f34b5ed422ea9/pkg/kv/kvserver/batcheval/cmd_lease_request.go#L99) to force a new lease sequence number - the increased start time leads to `!Equivalent()` [here](https://github.com/cockroachdb/cockroach/blob/ca60935df431525717592574475f34b5ed422ea9/pkg/kv/kvserver/batcheval/cmd_lease.go#L100). If we wanted to detect this case specifically from the lease request, we could compare `PrevLease.Start` to `MinProposedTS`.

But neither of these definitions help classify when a lease request is safe from getting wedged or not. In either case, it would be possible for the leaseholder to have been partitioned, replaced non-cooperatively, and fallen far behind on its log. What it sounds like you're getting at is that to feel sufficiently comfortable ignoring the lease-request-on-follower check, we'd want to make sure that the leaseholder couldn't have lost its lease non-cooperatively to some other replica. This gives us a reasonable assurance that the proposing replica doesn't have a wall of log entries waiting for it behind a network partition that it will need to wait on before hearing back about its lease request.

If that's the case, then maybe we shouldn't be worrying about the request and whether it looks like an extension at all. Maybe we should just be checking `(*Replica).ownsValidLeaseRLocked` in `FlushLockedWithRaftGroup` to determine whether we should allow a lease request to proceed on a follower.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> Please make the commit message explicit about the fact that the change is only about expiration-based leases

It's about epoch-based leases too. The initial suggestion was to forward all lease extensions (as defined by the previous and current lease having the same store ID in the lease request). However, this broke `TestRequestsOnLaggingReplica` since an epoch-based lease was being reclaimed by an old leaseholder in a new epoch, which was considered an extension, blocking requests to the old leaseholder.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-:-MVv0a1d-olEATj-2mCM:b-pkdiy5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 581 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVqcZ_E2DhhAre53RJO:-MVv1Wn5A7xc19hnh0PO:b-477djj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf.go#L581)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: `4` is to high, generally only used for extremely verbose messages in loops. Use the more usual `2`.
</blockquote></details>

Sure. I figured this was a fairly benign scenario compared to the others, and might be unnecessarily noisy since it'd be output on every lease extension, but will bump it to 2.
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVqhM9V9XC90lmRHM4_:-MVv1lmj-WzUxqvsrU4X:bhy9ot6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf_test.go#L269)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1 for `false /* newLeaseExtension */`. Please don't introduce an enum for a test helper function that's called 4 times, all in the same file.
</blockquote></details>

Will do.
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 591 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVtZpy57xEQrj9OYXSx:-MVv1qtn1DI7v7kjY3sv:bg2j5hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf_test.go#L591)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This name seems a little off. The follower is the one who holds the lease and is performing an extension, right? So should it be `follower, lease extension despite known eligible leader` or something along those lines?
</blockquote></details>

Makes sense, thanks.
___
*[pkg/roachpb/api.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVlf7r6CnBAHRSWFNkZ:-MVv2CGQD6xL3OGXOWCS:bikcpls) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/roachpb/api.go#L281)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is an interesting dilemma, as it strikes at the question of what constitutes a lease transfer and what lease requests are safe from getting wedged. I don't think the two sets are equivalent.

We consider a lease "extension" at different parts of the code to be one that [does not change the lease's owner](https://github.com/cockroachdb/cockroach/blob/ca60935df431525717592574475f34b5ed422ea9/pkg/kv/kvserver/batcheval/cmd_lease_request.go#L79). We make no attempt to say that the existing lease is still valid - if it is not and has already been replaced, the lease extension will fail.

An argument could be made that a lease that is replaced by a new lease with the same owner but a new sequence is not a true "extension", because writes proposed under the old lease will be rejected if reordered and applying after the new lease. This is the case during a node restart, where the node's `minLeaseProposedTS` is bumped, triggering [this logic](https://github.com/cockroachdb/cockroach/blob/ca60935df431525717592574475f34b5ed422ea9/pkg/kv/kvserver/batcheval/cmd_lease_request.go#L99) to force a new lease sequence number - the increased start time leads to `!Equivalent()` [here](https://github.com/cockroachdb/cockroach/blob/ca60935df431525717592574475f34b5ed422ea9/pkg/kv/kvserver/batcheval/cmd_lease.go#L100). If we wanted to detect this case specifically from the lease request, we could compare `PrevLease.Start` to `MinProposedTS`.

But neither of these definitions help classify when a lease request is safe from getting wedged or not. In either case, it would be possible for the leaseholder to have been partitioned, replaced non-cooperatively, and fallen far behind on its log. What it sounds like you're getting at is that to feel sufficiently comfortable ignoring the lease-request-on-follower check, we'd want to make sure that the leaseholder couldn't have lost its lease non-cooperatively to some other replica. This gives us a reasonable assurance that the proposing replica doesn't have a wall of log entries waiting for it behind a network partition that it will need to wait on before hearing back about its lease request.

If that's the case, then maybe we shouldn't be worrying about the request and whether it looks like an extension at all. Maybe we should just be checking `(*Replica).ownsValidLeaseRLocked` in `FlushLockedWithRaftGroup` to determine whether we should allow a lease request to proceed on a follower.
</blockquote></details>

Yep, great summary. I think `ownsValidLeaseRLocked()` might be the right solution here, but I wasn't sure if we could sometimes be proxying lease requests from other replicas or if the proposal buffer was exclusively for proposals the node generated itself. I'll try implementing this.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[pkg/roachpb/api.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVlf7r6CnBAHRSWFNkZ:-MVvUYAv9Cdo8ZknyxNA:b-x1cj6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/roachpb/api.go#L281)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Yep, great summary. I think `ownsValidLeaseRLocked()` might be the right solution here, but I wasn't sure if we could sometimes be proxying lease requests from other replicas or if the proposal buffer was exclusively for proposals the node generated itself. I'll try implementing this.
</blockquote></details>

Updated the PR, let me know what you think.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> Please make the commit message explicit about the fact that the change is only about expiration-based leases
> 
> > It's about epoch-based leases too. The initial suggestion was to forward all lease extensions (as defined by the previous and current lease having the same store ID in the lease request). However, this broke TestRequestsOnLaggingReplica since an epoch-based lease was being reclaimed by an old leaseholder in a new epoch, which was considered an extension, blocking requests to the old leaseholder.


But we've now established that "as defined by the previous and current lease having the same store ID" is the wrong definition for an extension, right? So what remains only affects expiration-based leases, I think? Like, there's no such thing as "an extension" for epoch-based leases, is there? For epoch-based ranges, a replica with a valid lease never proposes a new lease (does it?).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-:-MVxHne-6gSbySeIkMAP:b8nuamy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 567 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVxIqod8VWvYBGquSo-:-MVxIqod8VWvYBGquSo0:b5rz5cg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1beb8c63fd1980ca809206586b85bde8b6ce8388/pkg/kv/kvserver/replica_proposal_buf.go#L567)):*
> ```Go
> 		// lease. In that case, we have no choice but to continue with the proposal.
> 		//
> 		// Lease extensions for a currently held lease always go through.
> ```

please explain why they always go through in this comment
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 571 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVxIb405sQpeI3wrsAP:-MVxIb405sQpeI3wrsAQ:b-awotki) ([raw file](https://github.com/cockroachdb/cockroach/blob/1beb8c63fd1980ca809206586b85bde8b6ce8388/pkg/kv/kvserver/replica_proposal_buf.go#L571)):*
> ```Go
> 			req := p.Request.Requests[0].GetRequestLease()
> 			leaderKnownAndEligible := leaderInfo.leaderKnown && leaderInfo.leaderEligibleForLease
> 			isValidExtension := req.PrevLease.Replica.StoreID == req.Lease.Replica.StoreID &&
> ```

nit: I'd strike the "valid" from `isValidExtension`. What's an "invalid extension"?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 571 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVxJxCD5GGQI4JkCx19:-MVxJxCD5GGQI4JkCx1A:b-xeh8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/1beb8c63fd1980ca809206586b85bde8b6ce8388/pkg/kv/kvserver/replica_proposal_buf.go#L571)):*
> ```Go
> 			req := p.Request.Requests[0].GetRequestLease()
> 			leaderKnownAndEligible := leaderInfo.leaderKnown && leaderInfo.leaderEligibleForLease
> 			isValidExtension := req.PrevLease.Replica.StoreID == req.Lease.Replica.StoreID &&
> ```

is the `req.PrevLease.Replica.StoreID == req.Lease.Replica.StoreID` part necessary? What would go wrong if we just removed it?
`RequestLeaseRequest` always asks for a lease for the sender of the respective request, not for some other rando (unless you know otherwise?).


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
Thanks @erikgrinaker!

I think we can also address a TODO in `setupLeaseTransferTest` that references this issue. But when doing so, please stress `TestLeaseExpirationBelowFutureTimeRequest` a bit to make sure we aren't introducing any flakiness.

```
make stress PKG=./pkg/kv/kvserver TESTS= TestLeaseExpirationBelowFutureTimeRequest
```

> Reviewed 2 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-:-MVymCgCExbT39h8a1gs:boe82fd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @erikgrinaker)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 581 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVqcZ_E2DhhAre53RJO:-MVylTpn8NlE_xpELhme:bt7h9e7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf.go#L581)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Sure. I figured this was a fairly benign scenario compared to the others, and might be unnecessarily noisy since it'd be output on every lease extension, but will bump it to 2.
</blockquote></details>

Should we explain why we're proposing like we do in other cases? `; we hold the current lease`
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 571 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVxJxCD5GGQI4JkCx19:-MVyl5Ea35Llvc84_OFX:bmx8wh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1beb8c63fd1980ca809206586b85bde8b6ce8388/pkg/kv/kvserver/replica_proposal_buf.go#L571)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is the `req.PrevLease.Replica.StoreID == req.Lease.Replica.StoreID` part necessary? What would go wrong if we just removed it?
`RequestLeaseRequest` always asks for a lease for the sender of the respective request, not for some other rando (unless you know otherwise?).


</blockquote></details>

I was thinking the same thing. I don't see why we still need to look at the request at all.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 571 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVxIb405sQpeI3wrsAP:-MVylHWZAVAZ0vHSi1l1:b-kh62oi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1beb8c63fd1980ca809206586b85bde8b6ce8388/pkg/kv/kvserver/replica_proposal_buf.go#L571)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I'd strike the "valid" from `isValidExtension`. What's an "invalid extension"?
</blockquote></details>

I'd also rename this to `ownsCurrentLease` or something along those lines to reflect the salient point here.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 1029 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVyll-P42YmM0oGtpkc:-MVyll-P42YmM0oGtpkd:b-21lzzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1beb8c63fd1980ca809206586b85bde8b6ce8388/pkg/kv/kvserver/replica_proposal_buf.go#L1029)):*
> ```Go
> }
> 
> func (rp *replicaProposer) ownsValidLeaseRLocked(ctx context.Context, now hlc.ClockTimestamp) bool {
> ```

nit: let's keep this below `leaderStatusRLocked` so that these method impls remain ordered.
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVqhM9V9XC90lmRHM4_:-MVyls9r4FJJB5_R8Wwu:bjyov6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf_test.go#L269)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Will do.
</blockquote></details>

Should we remove this now?
___
*[pkg/roachpb/api.go, line 281 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVlf7r6CnBAHRSWFNkZ:-MVym3fo7DVR6TCgd22D:b9bkkpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/roachpb/api.go#L281)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Updated the PR, let me know what you think.
</blockquote></details>

Looks great.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
TFTRs, I've addressed the comments.

I tried removing the `AllowLeaseRequestProposalsWhenNotLeader` knob, but this causes `TestLeaseExpirationBelowFutureTimeRequest` to fail in about 1% of runs with:

```
    client_replica_test.go:1114: 
        	Error Trace:	client_replica_test.go:1114
        	            				subtest.go:23
        	Error:      	Expected nil, but got: &roachpb.Error{deprecatedMessage:"[NotLeaseHolderError] refusing to acquire lease on follower; r37: replica (n2,s2):2 not lease holder; current lease is repl=(n1,s1):1 seq=0 start=0,0 exp=<nil>", deprecatedTransactionRestart:0, UnexposedTxn:(*roachpb.Transaction)(nil), OriginNode:0, deprecatedDetail:roachpb.ErrorDetail{Value:(*roachpb.ErrorDetail_NotLeaseHolder)(0xc0005a4ce0)}, EncodedError:errorspb.EncodedError{Error:(*errorspb.EncodedError_Leaf)(0xc0005a4cc8)}, Index:(*roachpb.ErrPosition)(nil), Now:hlc.ClockTimestamp{WallTime:1616190382816199990, Logical:8, Synthetic:false}}
        	Test:       	TestLeaseExpirationBelowFutureTimeRequest/tooFarInFuture=false
```

I find this pretty odd, because if I call `replica1.GetLease()` it does indeed say that n2 is the leaseholder, and yet the request fails saying we're not. However, the current lease in the request above has `seq=0 start=0,0 exp=<nil>` -- could this be the initial lease, such that n2 somehow hasn't fully picked up the lease transfer yet?

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-:-MWS6gWpELmRACcJkJkh:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @erikgrinaker)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but consider clarifying the commit message to say that this patch is only about expiration-based leases (if you buy what I said above).

> I find this pretty odd, because if I call replica1.GetLease() it does indeed say that n2 is the leaseholder, and yet the request fails saying we're not. However, the current lease in the request above has seq=0 start=0,0 exp=<nil> -- could this be the initial lease, such that n2 somehow hasn't fully picked up the lease transfer yet?

The empty lease in the error makes me think that the replica that generated the error did not actually know the lease. Sometimes we generate `NotLeaseholderErrors` without having a lease on hand; I don't remember the exact cases.


Nit: it's convention to always respond with a "done" on the review comments after you've addressed them. Otherwise I don't know if they're still pending or not.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-:-MWUnNKh9ARGazOBPz8c:b-uqtsli)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andreimatei and @erikgrinaker)



<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> The empty lease in the error makes me think that the replica that generated the error did not actually know the lease. 

Ah, of course -- the error was from the other replica. Removed the test knob and fixed  `TestLeaseExpirationBelowFutureTimeRequest` by waiting for both replicas to pick up on the lease transfer.

> it's convention to always respond with a "done" on the review comments after you've addressed them.

Got it, will do! Still getting used to Reviewable.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-:-MWVmpJZE5eu-ZxlMogH:bz9ym6n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei and @erikgrinaker)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 571 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVxJxCD5GGQI4JkCx19:-MWVeGy_-HrV0zj0ZUYm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1beb8c63fd1980ca809206586b85bde8b6ce8388/pkg/kv/kvserver/replica_proposal_buf.go#L571)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I was thinking the same thing. I don't see why we still need to look at the request at all.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61982#-MVqhM9V9XC90lmRHM4_:-MWVeJ9L7UmcXaCFeLcL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1239745745af4d20e9acf0064eb0e7ff6bd129c2/pkg/kv/kvserver/replica_proposal_buf_test.go#L269)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we remove this now?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
bors r=nvanbenschoten,andreimatei,tbg
